### PR TITLE
setting appEui and appKey in provision()

### PR DESCRIPTION
**Pull request by Nicolasdejean**
added provision() #50

### CODE REVIEW

**Comment by FokkeZB** [comment]
@johanstokking any idea how in `showStatus()` the read for devEui is what I expect but now in `provision()` it returns what the previous command `sys reset` is supposed to return and not `sys get hweui`?

**Comment by FokkeZB** [comment]
@johanstokking I tried to fix this by changing `reset()` to:

```c
void TheThingsNetwork::reset(bool adr, int sf, int fsb) {
  #if !TTN_ADR_SUPPORTED
    adr = false;
  #endif

  modemStream->println(F("sys reset"));
  String version = readLine(3000);
  if (version == "") {
    debugPrintLn(F("Invalid version"));
    return;
  }

  model = version.substring(0, version.indexOf(' '));
  debugPrint(F("Version is "));
  debugPrint(version);
  debugPrint(F(", model is "));
  debugPrintLn(model);

  String devEui = readValue(F("sys get hweui"));
  String str = "";
  str.concat(F("mac set deveui "));
  str.concat(devEui);
  sendCommand(str);

  str = "";
  str.concat(F("mac set adr "));
  if(adr){
    str.concat(F("on"));
  } else {
    str.concat(F("off"));
  }
  sendCommand(str);
  // ..
```

But now I get an `ok` in response to `sys get hweui`:

```
-- STATUS
EUI: 0004A30B001B7AD2
Battery: 3213
AppEUI: 70B3D57EF0000028
DevEUI: 0004A30B001B7AD2
DevAddr: 00000000
Data Rate: 5
RX Delay 1: 1000
RX Delay 2: 2000
-- JOIN
Version is RN2483 1.0.1 Dec 15 2015 09:38:09, model is RN2483
Sending: mac set deveui 0004A30B001B7AD2
Sending: mac set adr off
Sending: mac set pwridx 1
Sending: mac set dr 5
Sending: mac set appeui with 8 bytes
Sending: mac set appkey with 16 bytes
Sending: mac save
Wait for OK time-out expired
Sending: mac set deveui ok
Response is not OK: invalid_param
```

Any idea?

**Comment by johanstokking** [comment]
@FokkeZB the `mac save` takes a while, it times out (see message). You should enlarge the time-out. The `ok` you get is a late response from `mac save`, the result of `sys get hweui` follows after.

Best is to debug these things with a serial passthrough.

File: src/TheThingsNetwork.cpp [hunk 1]
```diff
@@ -235,7 +237,8 @@ bool TheThingsNetwork::personalize() {
 bool TheThingsNetwork::provision(const byte appEui[8], const byte appKey[16]) {
   sendCommand(F("mac set appeui"), appEui, 8);
   sendCommand(F("mac set appkey"), appKey, 16);
-  sendCommand(F("mac save")); 
+  sendCommand(F("mac save"));
+  return waitForOK();
```
Comment (johanstokking) [inline src/TheThingsNetwork.cpp]: Expect a long wait here

File: API.md [hunk 2]
```diff
@@ -120,3 +120,16 @@ Calls `sendBytes()` with `{ 0x00 }` as payload to poll for incoming messages.
...
+
+```c
+bool provision(const byte appEui[8], const appKey[16]);
+```
+
+- `const byte appEui[8]`: Application Identifier for the device.
+- `const byte appKey[16]`: Application Key assigned to the device.
+
+Set the AppKey and AppEui in the beginning of the exemples.
+see the [Receive](https://github.com/TheThingsNetwork/arduino-device-lib/blob/master/examples/Receive/Receive.ino) exemple.
```
Comment (FokkeZB) [inline API.md]: The Receive examples doesn't have and will not have a `provision()`, so just leave this out.

File: API.md [hunk 3]
```diff
@@ -120,3 +120,16 @@ Calls `sendBytes()` with `{ 0x00 }` as payload to poll for incoming messages.
...
+Sets the information needed to activate the device.
+
+```c
+bool provision(const byte appEui[8], const appKey[16]);
+```
+
+- `const byte appEui[8]`: Application Identifier for the device.
+- `const byte appKey[16]`: Application Key assigned to the device.
+
+Set the AppKey and AppEui in the beginning of the exemples.
```
Comment (FokkeZB) [inline API.md]: Doesn't add much info, just drop the line

File: API.md [hunk 4]
```diff
@@ -120,3 +120,16 @@ Calls `sendBytes()` with `{ 0x00 }` as payload to poll for incoming messages.
 Returns the result of `sendBytes()`.
 
 See the [Receive](https://github.com/TheThingsNetwork/arduino-device-lib/blob/master/examples/Receive/Receive.ino) example.
+
+## Method: provision
+Sets the information needed to activate the device.
```
Comment (FokkeZB) [inline API.md]: ".. the device via OTAA, without actually activating. Call `join()` without the first 2 arguments to activate."

File: keywords.txt [hunk 5]
```diff
@@ -18,6 +18,7 @@ TTN_HEX_CHAR_TO_NIBBLE  KEYWORD2
 init                    KEYWORD2
 onMessage               KEYWORD2
 reset                   KEYWORD2
+provision		KEYWORD2
```
Comment (FokkeZB) [inline keywords.txt]: missing 1 space

File: src/TheThingsNetwork.cpp (hunk 1 above)
Comment (FokkeZB) [inline src/TheThingsNetwork.cpp]: Yes, as it is now it still times out and returns `ok` for the next command:

```c
Sending: mac save
Wait for OK time-out expired
Sending: mac set deveui ok
Response is not OK: invalid_param
Sending: mac join otaa
Join not accepted: denied
```

File: src/TheThingsNetwork.cpp [hunk 6]
```diff
@@ -237,8 +236,10 @@ bool TheThingsNetwork::personalize() {
 bool TheThingsNetwork::provision(const byte appEui[8], const byte appKey[16]) {
   sendCommand(F("mac set appeui"), appEui, 8);
   sendCommand(F("mac set appkey"), appKey, 16);
-  sendCommand(F("mac save"));
-  return waitForOK();
+  while (!sendCommand(F("mac save"))) {
+    delay(1000);
+  }
+  return waitForOK(10000);
```
Comment (FokkeZB) [inline src/TheThingsNetwork.cpp]: Did you test these new changes? I no longer get `ok` for `deveui` but for some reason still fails:

```
Sending: mac set appeui with 8 bytes
Sending: mac set appkey with 16 bytes
Sending: mac save
Wait for OK time-out expired
Sending: mac save
Sending: mac set deveui 0004A30B001B7AD2
Sending: mac join otaa
Join not accepted: denied
```

Comment (johanstokking) [inline src/TheThingsNetwork.cpp]: `sendCommand()` shouldn't take long. Also, with the `while` loop, your repeating sending the command.

Just one `sendCommand()` and a long wait in `waitForOK()`

Comment (Nicolasdejean) [inline src/TheThingsNetwork.cpp]: @FokkeZB I tested it join() works for me, that is what I got: 
```
Version is RN2483 1.0.1 Dec 15 2015 09:38:09, model is RN2483
Sending: mac set deveui 0004A30B001B4C9F
Sending: mac set adr off
Sending: mac set pwridx 1
Sending: mac set dr 5
Sending: mac set appeui with 8 bytes
Sending: mac set appkey with 16 bytes
Sending: mac save
Wait for OK time-out expired
Sending: mac save
Sending: mac set deveui 0004A30B001B4C9F
Sending: mac join otaa
Join accepted. Status: 00000401
EUI: 0004A30B001B4C9F
Battery: 3315
AppEUI: 70B3D57ED0000D35
DevEUI: 0004A30B001B4C9F
DevAddr: 1D1FF5E3
Band: 868
Data Rate: 5
RX Delay 1: 1000
RX Delay 2: 2000
```
For some reason ``` mac save ``` doesn't work the first time.

Comment (FokkeZB) [inline src/TheThingsNetwork.cpp]: I succeeded now, but still weird dat mac save times out, even if I use 10000 wait time:

```
Sending: mac set appeui with 8 bytes
Sending: mac set appkey with 16 bytes
Sending: mac save
Wait for OK time-out expired
Sending: mac set deveui 0004A30B001B7AD2
Sending: mac join otaa
Join accepted. Status: 00000401
```

**Comment by FokkeZB** [comment]
@johanstokking it works, but I do get a timeout after `mac save` still.

**Comment by Nicolasdejean** [comment]
@FokkeZB I put a higher wait-time for sendCommand, it works for me, try if it works for you

**Comment by FokkeZB** [comment]
Yes, that fixes it! @johanstokking good to merge

**Pull request merged**